### ISSUE TITLE
Some product information are lost or altered using a simple ObjectModel save

### ISSUE THREAD

**Issue by prestamodule** [PrestaShop/PrestaShop#17560]
**Describe the bug**
When using `save()` method on a product object, there are some issues

**To Reproduce**
Into a module, or anywhere in a file...
Use this simple code snippet:

```php
$product = new Product(1);
$product->save();
```

After the save(), unit price value displayed into BackOffice is reset to 0.

Now, if you are using this code snippet:

```php
// Let's use true as 2nd parameter to retrieve full product information
$product = new Product(1, true);
$product->save();
```

After the save():
* unit price value displayed into BackOffice is reset to 0.
* price without taxes is changed (reduced), each time a `save()` is done

**Comment by khouloudbelguith** [comment]
Hi @prestamodule,

Thanks for your report.
I manage to reproduce the issue with PS1.7.6.3 & PS1.7.5.2.
https://drive.google.com/file/d/1oGN0-ycdvMiu5wCUmR7YjzJtLd61KYjr/view
I used this script
```
<?php
require dirname(__FILE__).'/config/config.inc.php';
$product = new Product();
$product->id = 1;
$product->active = 0;
$product->name = array((int)Configuration::get('PS_LANG_DEFAULT') => 'testing');
$product->link_rewrite = array((int)Configuration::get('PS_LANG_DEFAULT') =>  'test');
$force_ids = true;
//this will force ObjectModel to use your ID
$product->force_id = (bool) $force_ids;
$product->save();
echo "Done, pro well updated";
?>
```
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

Related to: https://github.com/PrestaShop/PrestaShop/issues/17449

**Comment by Hlavtox** [comment]
Current behavior on develop.
- With ecotax disabled, everything seems to stay the same. ✅ 
- With ecotax enabled, ecotax is added to product price on every ObjectModel save. 🔴

### Why
In ProductCore __construct, there is this line..
`$this->price = Product::getPriceStatic((int) $this->id, false, null, 6, null, false, true, 1, false, null, null, null, $this->specificPrice);`
then when you check `getPriceStatic`, there is `$with_ecotax = true,` parameter, which causes ecotax to be added to price.

Correct solution I see is to gradually remove everything from the constructor, so the object model really serves only as a data storage.

Ping @kpodemski

**Comment by hibatallahAouadni** [comment]
Ping @Hlavtox & @kpodemski did you reproduce this issue with 81x?

**Comment by boo-code** [comment]
Answering @hibatallahAouadni's question: yes, it still reproduces, on 9.2.x.

Measured with `price = 10.00`, `ecotax = 2.00`, `unit_price = 5.00`, running `new Product($id, true); $p->save();` three times in a row:

| `PS_USE_ECOTAX` | ecotax | start | save 1 | save 2 | save 3 |
|---|---|---|---|---|---|
| 0 | 2.00 | 10.00 | **12.00** | **14.00** | **16.00** |
| 1 | 2.00 | 10.00 | **12.00** | **14.00** | **16.00** |
| 1 | 0.00 | 10.00 | 10.00 | 10.00 | 10.00 |
| 0 | 0.00 | 10.00 | 10.00 | 10.00 | 10.00 |

The zero-ecotax rows are the control: the price is stable, so the drift is exactly the ecotax.

One correction to @Hlavtox's 2023 note: it is **not** limited to ecotax being enabled. `getPriceStatic()` is called with `$with_ecotax = true` regardless of `PS_USE_ECOTAX`, so the first row drifts too. `unit_price_ratio` drifts with the price, because `updateUnitRatio()` recomputes from the row that was just corrupted.

The mechanism is the one @Hlavtox identified. `Product::__construct()` with `$full = true` keeps the stored price in `$base_price` and then overwrites `$price` with the computed final price; `ObjectModel::update()` persists `$price`, and the next load computes on top of the value it just wrote.

A narrow fix that does not wait for the constructor to be emptied: have `update()` write `$base_price` when `$price` still holds exactly what the constructor put there, and restore the computed value on the object afterwards. A price the caller actually assigned differs from the constructor's value and is saved as asked, and a product loaded without `$full` never gets a computed price at all. Nothing in core is affected either way — only four places build a product with `$full = true` (`ProductController`, `CartController`, `Pack`, `UpdateProductQuantityInCartHandler`) and all four only read.

Verified with an integration test covering the drift, the ratio, a caller-assigned price and a partially loaded product; reverting `classes/Product.php` fails the first two. `tests/Integration/Classes` is 187 green, and Behat `-s product` is 414 scenarios with one failure that fails identically on `upstream/9.2.x` on its own (`Attempt to read property "id" on null in ProductSearchContext.php:52`).

**Comment by Hlavtox** [comment]
@boo-code I think we should remove the price calculations from constructor, and use ObjectModel as data storage only.

**Comment by boo-code** [comment]
Agreed, and the measurement above says that is reachable rather than aspirational: only four places in core construct a product with `$full = true` — `ProductController`, `CartController`, `Pack` and `UpdateProductQuantityInCartHandler` — and all four only read the computed price. Nothing in core writes back a product it loaded that way, which is why the drift only shows up when someone calls `save()` on one.

So the destination is a constructor that loads columns and a caller that asks the price engine for a price. The part that needs deciding is not core but the compatibility surface: `$product->price` after a `$full = true` load is a computed, tax-and-specific-price-applied value that modules and themes read today, so emptying the constructor changes what that property means for them. That is a major-version change under ADR 0017 rather than something to slip into 9.2.x.

The narrow fix I described does not stand in the way of it: it only stops `update()` persisting the constructor's own computed value, so `save()` becomes idempotent while `price` keeps meaning exactly what it means now. If the constructor is emptied later, that guard becomes dead code and is deleted in the same commit.

I have it written with the integration test, against 9.2.x. Say which you would rather have and I will put it up — the guard now, or nothing here while the larger change is planned.
